third_party/accessibility/ax/platform/ax_platform_node_textrangeprovider_win.cc

```diff
@@ -0,0 +1,1701 @@
+
+#include "ax/ax_action_data.h"
+#include "ax/ax_range.h"
+#include "ax/platform/ax_platform_node_delegate.h"
+#include "ax/platform/ax_platform_node_win.h"
+#include "ax/platform/ax_platform_tree_manager.h"
+#include "base/win/variant_vector.h"
+
+#define UIA_VALIDATE_TEXTRANGEPROVIDER_CALL()                  \
+  if (!GetOwner() || !GetOwner()->GetDelegate() || !start() || \
+      !start()->GetAnchor() || !end() || !end()->GetAnchor())  \
+    return UIA_E_ELEMENTNOTAVAILABLE;                          \
+  SetStart(start()->AsValidPosition());                        \
+  SetEnd(end()->AsValidPosition());
+#define UIA_VALIDATE_TEXTRANGEPROVIDER_CALL_1_IN(in)           \
+  if (!GetOwner() || !GetOwner()->GetDelegate() || !start() || \
+      !start()->GetAnchor() || !end() || !end()->GetAnchor())  \
+    return UIA_E_ELEMENTNOTAVAILABLE;                          \
+  if (!in)                                                     \
+    return E_POINTER;                                          \
+  SetStart(start()->AsValidPosition());                        \
+  SetEnd(end()->AsValidPosition());
+#define UIA_VALIDATE_TEXTRANGEPROVIDER_CALL_1_OUT(out)         \
+  if (!GetOwner() || !GetOwner()->GetDelegate() || !start() || \
+      !start()->GetAnchor() || !end() || !end()->GetAnchor())  \
+    return UIA_E_ELEMENTNOTAVAILABLE;                          \
+  if (!out)                                                    \
+    return E_POINTER;                                          \
+  *out = {};                                                   \
+  SetStart(start()->AsValidPosition());                        \
+  SetEnd(end()->AsValidPosition());
+#define UIA_VALIDATE_TEXTRANGEPROVIDER_CALL_1_IN_1_OUT(in, out) \
+  if (!GetOwner() || !GetOwner()->GetDelegate() || !start() ||  \
+      !start()->GetAnchor() || !end() || !end()->GetAnchor())   \
+    return UIA_E_ELEMENTNOTAVAILABLE;                           \
+  if (!in || !out)                                              \
+    return E_POINTER;                                           \
+  *out = {};                                                    \
+  SetStart(start()->AsValidPosition());                         \
+  SetEnd(end()->AsValidPosition());
+// Validate bounds calculated by AXPlatformNodeDelegate. Degenerate bounds
+// indicate the interface is not yet supported on the platform.
+#define UIA_VALIDATE_BOUNDS(bounds)                           \
+  if (bounds.OffsetFromOrigin().IsZero() && bounds.IsEmpty()) \
+    return UIA_E_NOTSUPPORTED;
+
+namespace ui {
+
+class AXRangePhysicalPixelRectDelegate : public AXRangeRectDelegate {
+ public:
+  explicit AXRangePhysicalPixelRectDelegate(
+      AXPlatformNodeTextRangeProviderWin* host)
+      : host_(host) {}
+
+  gfx::Rect GetInnerTextRangeBoundsRect(
+      AXTreeID tree_id,
+      AXNode::AXID node_id,
+      int start_offset,
+      int end_offset,
+      AXOffscreenResult* offscreen_result) override {
```

Comment:
Should we add the `clipping_behavior` parameter to `AXRangeRectDelegate`?

Reply:
Probably. It does not look like we actually use the class outside of this file

Reply:
I will put it to a TODO for now.